File: common/src/main/java/generations/gg/generations/core/generationscore/network/packets/shop/S2COpenShopHandler.java
package generations.gg.generations.core.generationscore.network.packets.shop;

import generations.gg.generations.core.generationscore.client.screen.ShopScreen;
import generations.gg.generations.core.generationscore.network.ClientNetworkPacketHandler;
import generations.gg.generations.core.generationscore.world.entity.ShopOfferProvider;
import net.minecraft.client.Minecraft;

public class S2COpenShopHandler implements ClientNetworkPacketHandler<S2COpenShopPacket> {

    @Override
    public void handle(S2COpenShopPacket packet) {
        if (Minecraft.getInstance().level == null)
            return;

        if (packet.entityId != -1 && Minecraft.getInstance().level.getEntity(packet.entityId) instanceof ShopOfferProvider shopOfferProvider)
            Minecraft.getInstance().setScreen(new ShopScreen(shopOfferProvider));
        else if (Minecraft.getInstance().level.getBlockEntity(packet.pos) instanceof ShopOfferProvider shopOfferProvider) {
            Minecraft.getInstance().setScreen(new ShopScreen(shopOfferProvider));
        }
    }
}

File: src/common/main/kotlin/generations/gg/generations/core/generationscore/common/network/packets/shop/S2COpenShopHandler.kt
package generations.gg.generations.core.generationscore.common.network.packets.shop

import generations.gg.generations.core.generationscore.common.client.screen.ShopScreen
import generations.gg.generations.core.generationscore.common.world.entity.ShopOfferProvider
import net.minecraft.client.Minecraft

object S2COpenShopHandler : com.cobblemon.mod.common.api.net.ClientNetworkPacketHandler<S2COpenShopPacket> {
    override fun handle(packet: S2COpenShopPacket, client: Minecraft) {
        if (Minecraft.getInstance().level == null) return

        val entity = Minecraft.getInstance().level!!.getEntity(packet.entityId)

        if (packet.entityId != -1 && entity is ShopOfferProvider) Minecraft.getInstance().setScreen(ShopScreen(entity))
        else {
            val blockEntity = Minecraft.getInstance().level!!.getBlockEntity(packet.pos!!)

            if (blockEntity is ShopOfferProvider) {
                Minecraft.getInstance().execute { Minecraft.getInstance().setScreen(ShopScreen(blockEntity)) }
            }
        }
    }
}
